### PR TITLE
Stripe integration frontend

### PR DESCRIPTION
**Pull request by MauAraujo**
## What does this PR do?
This PR adds all frontend changes related to the Stripe integration. The most important changes are:
- Rework the Billing tab shown under project settings
- Add the `BillingModal` that can be reused to add payment methods
- Add the Stripe `PaymentElement` to collect payment details
- Add api routes that correspons to the ones created in #4430 
- Add stripe publishable key to configuration

## Implementation
I decided to use the Stripe  library since it has some pre-made elements that take care of validation. They also handle redirects to specific credit card confirmation pages, and it allows the page to be compliant. It's recommended by Stripe to use those pre-made elements.

## Screenshots

https://github.com/porter-dev/porter/assets/19579265/020f163d-bcc9-48cc-8b05-71b947521fbc

### CODE REVIEW

File: dashboard/src/main/home/modals/PaymentSetupForm.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            setErrorMessage(error.message);
+        }
+
+        onCreate()
+    };
+
+    return (
+        <form>
+            <PaymentElement />
+            <SubmitButton className='submit-button' text={"Add Payment Method"} disabled={!stripe || loading} onClick={handleSubmit}>
```
Comment (ferozemohideen) [inline dashboard/src/main/home/modals/PaymentSetupForm.tsx:51]: we have a Button component in `components/porter`, that will allow you to pass in an error message

File: dashboard/src/main/home/project-settings/BillingPage.tsx [hunk 2]
```diff
@@ -1,69 +1,170 @@
...
-      queryUsage();
-    };
-  }, [currentProject?.id]);
+    (async () => {
+      await api.checkBillingCustomerExists("<token>", { user_email: user?.email }, { project_id: currentProject?.id })
+      const listResponse = await api.listPaymentMethod("<token>", {}, { project_id: currentProject?.id })
+      const paymentMethodList = listResponse.data === null ? [] : listResponse.data
+      setPaymentMethods(paymentMethodList)
+    })();
+  }, []);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/project-settings/BillingPage.tsx:30]: can replace this with a `useQuery`. See how it's used in one of the `/lib/hooks/` files for an example, basically handles data fetching and tracking loading states for you

File: dashboard/src/main/home/project-settings/BillingPage.tsx [hunk 3]
```diff
@@ -1,69 +1,170 @@
...
+
+  const deletePaymentMethod = async (paymentMethod) => {
+    setDeleteStatus(true)
+    const resp = await api.deletePaymentMethod("<token>", {}, { project_id: currentProject?.id, payment_method_id: paymentMethod.id })
+    if (resp.status !== 200) {
+      return <Error message="failed to delete payment method" />
+    }
+
+    setDeleteStatus(false)
+    setPaymentMethods(paymentMethods.filter(elem => elem !== paymentMethod))
```
Comment (ferozemohideen) [inline dashboard/src/main/home/project-settings/BillingPage.tsx:49]: I would put these methods in a file called `/lib/hooks/useStripe.ts` to separate the logic from component styling

File: dashboard/src/main/home/project-settings/BillingPage.tsx [hunk 4]
```diff
@@ -1,69 +1,170 @@
...
-              },
-            },
-          }}
-        ></PlanSelect>
-      </CustomerProvider>
+      <BillingModalWrapper>
+        <Heading isAtTop={true}>Payment methods</Heading>
+        <Helper>
+          This displays all configured payment methods
+        </Helper>
```
Comment (ferozemohideen) [inline dashboard/src/main/home/project-settings/BillingPage.tsx:68]: we use the `Text` component in `components/porter`, with color='helper' attribute

File: dashboard/src/main/home/project-settings/BillingPage.tsx [hunk 5]
```diff
@@ -1,69 +1,170 @@
...
+import Error from "components/porter/Error";
+import Loading from "components/Loading";
 
 function BillingPage() {
-  const [customerToken, setCustomerToken] = useState("");
-  const [teamID, setTeamID] = useState("");
-  const { currentProject, setCurrentError, queryUsage } = useContext(Context);
+  const { user, currentProject } = useContext(Context);
+
+  const [paymentMethods, setPaymentMethods] = useState([]);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/project-settings/BillingPage.tsx:19]: this should be typed

**Review by ferozemohideen — CHANGES_REQUESTED** [review]
Overall this is a good start - would recommend using `useQuery` and `zod` validation for backend calls and also you should turn `simplified_view_enabled` on to see the new dashboard style

File: dashboard/src/main/home/project-settings/BillingPage.tsx (hunk 2 above)
Comment (MauAraujo) [reply dashboard/src/main/home/project-settings/BillingPage.tsx]: Ok, I moved all Stripe specific logic to  `/lib/hooks` and rewrote to use   `useQuery`

File: dashboard/src/main/home/project-settings/BillingPage.tsx (hunk 3 above)
Comment (MauAraujo) [reply dashboard/src/main/home/project-settings/BillingPage.tsx]: Thanks for the suggestion, moved in latest commit

File: dashboard/src/main/home/project-settings/BillingPage.tsx (hunk 5 above)
Comment (MauAraujo) [reply dashboard/src/main/home/project-settings/BillingPage.tsx]: Fixed!

File: dashboard/src/main/home/project-settings/BillingPage.tsx (hunk 4 above)
Comment (MauAraujo) [reply dashboard/src/main/home/project-settings/BillingPage.tsx]: Got it, changed to use `Text` instead of the `Helper` component

**Pull request closed without merging**